### PR TITLE
Set structure factor background default to 0.0. Refs #1157

### PR DESCRIPTION
**Pull request by pkienzle**
Since background/scale for structure factors is not available from sasview, make sure that background defaults to 0.0 if it is missing.

See [Ticket #1157](http://trac.sasview.org/ticket/1157).

### CODE REVIEW

File: sasmodels/modelinfo.py [hunk 1]
```diff
@@ -771,9 +778,24 @@ def make_model_info(kernel_module):
...
+
+    # Build the parameter table
     #print("make parameter table", kernel_module.parameters)
     parameters = make_parameter_table(getattr(kernel_module, 'parameters', []))
+
+    # background defaults to zero for structure factor models
+    structure_factor = getattr(kernel_module, 'structure_factor', False)
+    if structure_factor:
+        # Make sure background is the second common parameter
+        assert COMMON_PARAMETERS[1][0] == "background"
```
[the remark this answers is not in the document]
Comment (pkienzle) [reply sasmodels/modelinfo.py]: I intended that it be redundant.  Harking back to the days of Eiffel (the programming language) and programming by contract, you can improve safety using pre-conditions and post-conditions on code blocks.  But you are correct that a direct test of common_parameters is a better contract.

**Comment by krzywon** [comment]
I compared fits on the same set of data to the same model + structure factor using this branch of sasmodels and the master branch. I took the parameters from the fit in this branch and entered them in using master and compared the fits. The results are shown below.

fit_4_2_0_beta_10k_a_sphere_plus_squarewell_branch_ticket-1157
Figure 1: A plot of the dataset 10000A_sphere_dsm.xml in the test/1d_data folder, fit to a combined S(Q)*P(Q) of the sphere model and square well structure factor using sasmodels branch ticket-1175.
fit_4_2_0_beta_10k_a_sphere_plus_squarewell_branch_master
Figure 2: A plot of the dataset 10000A_sphere_dsm.xml in the test/1d_data folder to a combined S(Q)*P(Q) of the sphere model and square well structure factor using sasmodels master branch. Parameters for the fit were copied from the fit using in Figure 1.

The difference in the fits is subtle, but the master branch fit is shifted down slightly, due to the background subtraction from the structure factor that shouldn't be there.

There was a lot of talk in the ticket on what should be done in this case, so I am not merging yet, but the code does what it says it does.

**Comment by pkienzle** [comment]
Don't know what is going on here.  When I change modelinfo.py(791) to have a default background of 1000.0, I don't see any change for sphere@hardsphere in sascomp.

Can you redo your experiment changing the default in the new version, and see if you get the same result?

**Comment by butlerpd** [comment]
The discussion in the ticket is relevant to 4.3 or 5.0 but not 4.2.  In 4.2 neither scale nor background are accessible to the user and thus should be set to 1.0 and 0.0.  The discussion has to do with them being available to the user to alter and/or fit - an inappropriate level of change at this point of the release process.

However I am very confused by the plots.  the background in master as I understand it is set to 0.001 rather than 0.0. So firstly, I would think that for exactly the same parameters otherwise, master should yield a curve shifted slightly HIGHER than the branch.  More importantly however, a 0.001 difference should be completely invisible on a log scale at 10^8 (I0 above) and thus not in fact visible at all in these plots. Not sure I understand @pkienzle comment above but it sounds wrong as well.  We need to understand this behavior before we can merge I think.

**Comment by butlerpd** [comment]
How does this affect the sum|multiply models for which, in the multiply version, neither model should have a background (or scale as I recall) with a scale and background added to the product.  This was a huge bug in 4.0 which was specifically fixed by 4.1.2.

**Comment by butlerpd** [comment]
I am confused.  Tests of master (without this PR) indicate that in fact the default for background is forced to zero for S(Q) and 0.001 otherwise.  This suggests that wherever the specific model populating a tab has the background and scale suppressed ONLY for S(Q) must also be forcing the default background to zero at the same time.  Looking at the code that does indeed seem to be the case.  In sasmodels.sasview_model the following code is found
```
        if self._model_info.structure_factor:
            hidden.add('scale')
            hidden.add('background')
            self._model_info.parameters.defaults['background'] = 0.
```

The PR also sets `use-demo` to False instead of True (in preparation for removing all `demo` code?) and removes the demo dict for `hardsphere`  Is the point of the other changes to ensure zero is forced after sasview_model is not longer used? .. and when would that be?  It should not be a part of sasmodels in the long run I guess.  

That said the code looks fine and still works as before regarding background so could be merged.

Question: should this be a 4.2 change?

**Comment by butlerpd** [comment]
Note tests were:
* cylinder model (use low scale factor and large radius and length so you hit background faster but leave default background which is 0.001) - asymptote 0.00100
* hardsphere model (no scale and or background available but set vol frac very very low so oscillations are sufficiently damped by 0.5 1/A) - asymptote 0.00000
* sphere * hardsphere (set volume fraction low again but leave default background of 0.001 alone) - asymptote 0.00100

**Comment by pkienzle** [comment]
This PR sets background to 0 for the sasview wrapper to structure factors (scale is already 1).  It should be applied for 4.2; the user has no control over the background value since the sasview wrapper hides it from the user interface, so there is no work-around.

Changing use_demo to false and removing demo from hardsphere has nothing to do with the ticket; I just happened to be on this branch when I changed it.  SasView doesn't use demo, so the change is harmless, other than being confusing that it is in this PR.  Yes, it would have been better to put it on its own branch, along with purging all the demo=... attributes from the other models.

**Comment by butlerpd** [comment]
the confusion is that this "fix" does **absolutely nothing** from the perspective of the functionality from what I can tell - what is the use case where this patch behaves differently than master?

**Comment by pkienzle** [comment]
Apparently younger me already anticipated this problem.  The code you show above (from 2016-10-10) does indeed set background to zero in the sasview wrapper for structure factor models.

This PR serves only to set the default background to zero for structure factors when using sasmodels directly (e.g., via sascomp or the fitting examples), and can be held until 4.3.

**Comment by butlerpd** [comment]
Ultimately I believe we want to remove this remaining sasview wrapper from the sasmodels package - which would mean that in fact the forcing to zero would need to drop to the SasView package itself which this PR does.  The changes are small enough and functionality test shows no change in behavior so should be safe to merge however, after discussing with @pkienzle, being disciplined and professional we will wait to merge this till after the release.

**Pull request merged**